src/utl/src/timer.cpp

```diff
@@ -74,4 +78,71 @@ DebugScopedTimer::~DebugScopedTimer()
   debugPrint(logger_, tool_, group_.c_str(), level_, msg_, *this);
 }
 
+ScopedStatistics::ScopedStatistics(utl::Logger* logger, std::string msg)
+    : Timer(),
+      msg_(std::move(msg)),
+      start_rsz_(getStartRSZ()),
+      start_vsz_(getStartVSZ()),
+      logger_(logger)
+{
+}
+
+size_t ScopedStatistics::getMemoryUsage(const char* tag)
+{
+  FILE* file = fopen("/proc/self/status", "r");
+  if (file == nullptr) {
+    perror("Failed to open /proc/self/status");
+    exit(EXIT_FAILURE);
+  }
+
+  size_t result = (size_t) -1;
+  char line[128];
+  size_t tagLength = strlen(tag);
+
+  while (fgets(line, sizeof(line), file) != nullptr) {
+    if (strncmp(line, tag, tagLength) == 0) {
+      const char* p = line;
+      while (*p < '0' || *p > '9') {
+        p++;
+      }
+      result = (size_t) atoi(p);
+      break;
+    }
+  }
+
+  fclose(file);
+  if (result == (size_t) -1) {
+    fprintf(stderr, "%s not found in /proc/self/status\n", tag);
+    exit(EXIT_FAILURE);
+  }
+
+  return result / 1024;
+}
+
+size_t ScopedStatistics::getStartVSZ()
+{
+  return getMemoryUsage("VmSize:");
+}
+
+size_t ScopedStatistics::getPeakVSZ()
+{
+  return getMemoryUsage("VmPeak:");
+}
+
+size_t ScopedStatistics::getStartRSZ()
+{
+  return getMemoryUsage("VmRSS:");
+}
+
+size_t ScopedStatistics::getPeakRSZ()
+{
+  return getMemoryUsage("VmHWM:");
+}
+
+ScopedStatistics::~ScopedStatistics()
+{
+  logger_->report(msg_ + ": runtime {} seconds, usage: rsz = {} MB, vsz = {} MB, peak: rsz = {} MB, vsz = {} MB", static_cast<int>(Timer::elapsed()), (getPeakRSZ() - start_rsz_),
```

Comment:
runtime is unclear as it could be elapsed or cpu time.  I think using elapsed is clearer.

Personally I think peak is sufficient (@oharboe any opinion)?

Reply:
If we don't know the starting memory consumption, we can miss false positives: the problem is not with the algorithm that is reported as too high usage, but something in the setup.

Reply:
This doesn't tell you the starting memory - only the end - start

Reply:
correct. pick a way to present two numbers....

start memory is trivially calculated from peak and usagr, true for any cpmbination of these two numbers. most time usage and peak is what is of interested

Reply:
usage and peak do not give start.  The peak may well be in the middle not at the end.

Reply:
Now I am confused as to what this code is measuring. Seems like that should be clear before merging(if not to me, then everyone else....)

What I really care about is to know how much ram that I should have for a stage. I dont care what my machine has, I just.need to know what the correct amount of ram for the job is so I can make it so.

Reply:
Which is what peak should say and why I don't think usage matters much.

Reply:
Agreed. Keep it simple. If there is something wrong(large initial memory allocation outside of scope), then it can be debugged otherwise. I think normal operation, determining required RAM, is more important and other information clouds that...

Reply:
so let's go with the DRT format above (with prefix)

Reply:
From the conversation below it's not clear what the final message should be.  Please confirm if it should look like the DRT message you provided.
 `[INFO DRT-0267] cpu time = 00:00:04, elapsed time = 00:00:01, memory = 2197.27 (MB), peak = 4031.35 (MB)`
Otherwise please provide the exact way you would like the message to be.

Reply:
I was thinking the prefix would be about what is being timed, eg

`repair_design: cpu time = 00:00:04, elapsed time = 00:00:01, memory = 2197.27 (MB), peak = 4031.35 (MB)`

Though the prefix would be up to the client code.